### PR TITLE
Modeling Rules Testing

### PR DESCRIPTION
**Pull request by avidan-H**
## Status
- [x] Ready

## Related Issues
fixes: link to the issue

## Description
Implements two hidden commands:
- `demisto-sdk modeling-rules test`
- `demisto-sdk modeling-rules init-test-data` 

## Screenshots
`demisto-sdk modeling-rules test`
image

`demisto-sdk modeling-rules init-test-data` 
image

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

**Comment by avidan-H** [comment]
Doesn't have any tests yet but I want to get the commands into the SDK so that the security team can already begin using and testing it out so we can get feedback.

File: demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        False,
+        help='Quiet output - sets verbosity to default.',
+        rich_help_panel='Logging Configuration',
+    ),
+    log_path: Path = typer.Option(
+        None,
+        '-lp', '--log-path',
+        resolve_path=True,
+        show_default=False,
+        help='Path of directory in which you would like to store all levels of logs.',
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py:57]: Add:
If not given, then log_file_name will be disregarded, and the log output will be to stdout

File: demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+from demisto_sdk.commands.common.logger import setup_rich_logging
+from demisto_sdk.commands.test_content.xsiam_tools.test_data import (EventLog,
+                                                                     TestData)
+
+app = typer.Typer()
+logger = logging.getLogger('demisto-sdk')
+
+
+@app.command(no_args_is_help=True)
+def init_test_data(
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py:20]: Add documentation that says that this can be used for `create` or `update`.

File: demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                logger.info(f'[cyan]Updating test data file: {test_data_file}[/cyan]', extra={'markup': True})
+                test_data = TestData.parse_file(test_data_file)
+                for event_log in test_data.data:
+                    if not event_log.vendor:
+                        event_log.vendor = default_vendor
+                    if not event_log.product:
+                        event_log.product = default_product
+                    if not event_log.dataset:
+                        event_log.dataset = default_dataset
+                # for expected in test_data.expected_values:
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py:97]: Align the indentation or remove

File: demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                                keys_to_remove.append(key)
+                        for key in keys_to_remove:
+                            event_log.expected_values.pop(key)
+                        new_mapping.update(event_log.expected_values)
+
+                    event_log.expected_values = new_mapping
+
+                if count > len(test_data.data):
+                    # create the missing templated data and add it to the test data
+                    templated_event_data_to_add = [
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py:114]: Use itertools.repeat

File: demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                    ]
+
+                    test_data.data.extend(templated_event_data_to_add)
+            else:
+                logger.info(
+                    f'[cyan]Creating test data file for: {mr_entity.path.parent}[/cyan]',
+                    extra={'markup': True}
+                )
+                test_data = TestData(
+                    data=[
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py:130]: Use itertools.repeat

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+console = Console(theme=custom_theme)
+
+
+app = typer.Typer()
+
+
+ContentEntity = Union[YAMLContentUnifiedObject, YAMLContentObject, JSONContentObject]
+
+
+def create_table(expected: Dict[str, Any], received: Dict[str, Any]) -> Table:
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:52]: Consider moving to rich_utilities

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+            if str(e.test_data_event_id) == td_event_id:
+                expected_values = e.expected_values
+                break
+
+        if expected_values:
+            printr(create_table(expected_values, result))
+
+            for key, val in expected_values.items():
+                if not val:
+                    # TODO: Make this a debugging statement
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:104]: Remove

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+    """Generate an XQL query from the given rule and test data event IDs.
+
+    Args:
+        rule (MRule): Rule object parsed from the modeling rule file.
+        test_data_event_ids (List[str]): List of test data event IDs to query.
+
+    Returns:
+        str: The XQL query.
+    """
+    fields = ', '.join([f'{f}' for f in rule.fields])
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:141]: Check if `rule.fields` is strings

**Review by tkatzir — COMMENTED** [review]
Looks good!
Some comments.

File: demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py (hunk 4 above)
Comment (avidan-H) [reply demisto_sdk/commands/test_content/test_modeling_rule/init_test_data.py:114]: `itertools.repeat` repeats the same object multiple times versus instantiating multiple `EventData` instances which is actually the behavior I want.

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+    """Push the test data to the tenant.
+
+    Args:
+        xsiam_client (XsiamApiClient): Xsiam API client.
+        mr (ModelingRule): Modeling rule object parsed from the modeling rule file.
+        test_data (init_test_data.TestData): Test data object parsed from the test data file.
+    """
+    events_test_data = [e.event_data for e in test_data.data]
+    for i, event_log in enumerate(test_data.data):
+        if isinstance(event_log.event_data, dict):
```
[the remark this answers is not in the document]
Comment (avidan-H) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:214]: Actually, without the instance check, there is a potential type error since _technically_ `event_data` can be `None` according to how it is typed.

File: demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+            "request_data": {
+                "query": query
+            }
+        }
+        endpoint = urljoin(self.base_url, 'public_api/v1/xql/start_xql_query/')
+        logger.info(f'Starting xql query:\nendpoint={endpoint}\n{query=}')
+        response = self.session.post(endpoint, json=body)
+        data = response.json()
+
+        if 200 <= response.status_code < 300:
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py:239]: Standardize with `range()`

File: demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+            "request_data": {
+                "query_id": execution_id,
+                "pending_flag": False,
+                "limit": 1000,
+                "format": "json"
+            }
+        })
+        endpoint = urljoin(self.base_url, 'public_api/v1/xql/get_query_results/')
+        logger.info(f'Getting xql query results: endpoint={endpoint}')
+        response = self.session.post(endpoint, data=payload)
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py:259]: Can you set some timeout?

Comment (tkatzir) [inline demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py:259]: logger.debug('returned...')

File: demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+        data = response.json()
+        logger.debug(pformat(data))
+
+        if 200 <= response.status_code < 300 and data.get('reply', {}).get('status', '') == 'SUCCESS':
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py:263]: Standardize with `range()`

File: demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            reply_results_data = data.get('reply', {}).get('results', {}).get('data', [])
+            return reply_results_data
+        else:
+            err_msg = (f'Failed to get xql query results for execution_id "{execution_id}"'
+                       f' - with status code {response.status_code}\n{pformat(data)}')
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/xsiam_tools/xsiam_interface.py:268]:
```suggestion
                       f' - with status code {response.status_code}. data:\n{pformat(data)}')
```

File: demisto_sdk/commands/upload/upload.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import os
```
Comment (tkatzir) [inline demisto_sdk/commands/upload/upload.py:1]: Do in a separate PR

File: demisto_sdk/utils/utils.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
+import os
```
Comment (tkatzir) [inline demisto_sdk/utils/utils.py:1]: Separate PR

**Review by tkatzir — COMMENTED** [review]
We're getting close!

File: demisto_sdk/utils/utils.py (hunk 15 above)
Comment (avidan-H) [reply demisto_sdk/utils/utils.py:1]: See https://github.com/demisto/demisto-sdk/pull/2518

**Review by tkatzir — APPROVED** [review]
Great!
